File: Exponentiation.cpp
/* ==================================================
Exponentiation
Shikuan Huang
================================================== */

#include <iostream>
using namespace std;

// Basic recursive implementation
long long rExp(long long number, int power)
{	
	if (power == 0)
		return 1;
	
	return number * rExp(number, power - 1);
}

// Basic recursive implementation for negative powers
double rNExp(double number, int power)
{
	if (power == 0)
		return 1;
	
	return (1/number) * rNExp(number, power - 1);
}

// Tail recursive implementation
long long tExp(int number, int power, long long partialProduct = 1)
{
	if (power == 0)
		return partialProduct * 1;
	
	return tExp(number, power - 1, partialProduct * number);
}

// Iterative implementation
long long iExp(int number, int power)
{
	long long result = 1;
	while (power != 0)
	{
		result = result * number;
		power--;
	}
	return result;
}

// Calls the positive and negative versions depending on "power"
long double exp(int number, int power)
{
	if (power < 0)
		return rNExp(number, -power);
	else
		return iExp(number, power);	
}

int main()
{
	int base;
	int power;
	bool end = false;
	
	while (!end)
	{
		cout << "Enter a base and a power: ";
		cin >> base >> power;

		cout << base << "^" << power << " = " << exp(base, power) << endl << endl;
	}
}

File: Java/Exponentiation.java
/* ==================================================
 * Exponentiation
 * Shikuan Huang
 * 
 * Basic recursive implementation
 * Dynamic programming implementation
 * Tail recursive implementation
 * Iterative implementation
 * ==================================================
 */

import java.util.*;
import java.math.*;

public class Exponentiation {
  public static void main(String[] args) {
	Scanner input = new Scanner(System.in);
	System.out.print("Enter a base and a power: ");
	while (input.hasNext()) {
	  int inputBase = input.nextInt();
	  int inputPower = input.nextInt();
	  BigInteger base = new BigInteger(String.valueOf(inputBase));
	  BigInteger power = new BigInteger(String.valueOf(inputPower));
	  System.out.println(base + " raised to the power of " + power + " is " + iPow(base, power) + "\n");
	  System.out.print("Enter a base and a power: ");
	} 
	input.close();
  }
  
  // Basic recursive implementation
  static BigInteger rPow(BigInteger base, BigInteger power) {
	if (power.equals(BigInteger.ZERO)) {
	  return BigInteger.ONE; 
	}
	return base.multiply(rPow(base, power.subtract(BigInteger.ONE)));
  }
  
  // Dynamic programming implementation
  static BigInteger dPow(BigInteger base, BigInteger power) {
	BigInteger[] answers = new BigInteger[power.intValue() + 1];
	Arrays.fill(answers, new BigInteger(String.valueOf(-1)));
	return dPow(base, power, answers);
  }
  
  static BigInteger dPow(BigInteger base, BigInteger power, BigInteger[] answers) {
    BigInteger NegOne = new BigInteger(String.valueOf(-1));
	if (!answers[power.intValue()].equals(NegOne)) {
	  return answers[power.intValue()];
	}
	
	if (power.equals(BigInteger.ZERO)) {
	  answers[power.intValue()] = BigInteger.ONE;
	  return BigInteger.ONE;
	}
	
	answers[power.intValue()] = 
	    base.multiply(dPow(base, power.subtract(BigInteger.ONE), answers));
	return answers[power.intValue()];
  }
  
  // Tail recursive implementation
  static BigInteger tPow(BigInteger base, BigInteger power) {
	return tPow(base, power, BigInteger.ONE);
  }
  
  static BigInteger tPow(BigInteger base, BigInteger power, BigInteger partialProd) {
    if (power.equals(BigInteger.ZERO)) {
      return partialProd;
    }
    
    return tPow(base, power.subtract(BigInteger.ONE), partialProd.multiply(base));
  }
  
  // Iterative implementation
  static BigInteger iPow(BigInteger base, BigInteger power) {
	BigInteger partialProd = BigInteger.ONE;
	for (int i = 0; i < power.intValue(); i++) {
	  partialProd = partialProd.multiply(base);
	}
	return partialProd;
  }
}

